File: apps/user/src/functions/UserAuthorizer.ts
'use strict'

import {Handler} from "aws-lambda";

const AbstractAuthorizer = require('../../../../shared/common/authorization/AbstractAuthorizer')
const {AuthPolicy, HttpVerb} = require('../../../../shared/common/authorization/AuthPolicy')

class UserAuthorizer extends AbstractAuthorizer {
    generatePermissions(authPolicy: typeof AuthPolicy, userId: string) {
        if (userId) {
            authPolicy.allowMethod(HttpVerb.GET, `/v1/users/search`)
            authPolicy.allowMethod(HttpVerb.POST, `/v1/users/notify`)
            authPolicy.allowMethod(HttpVerb.GET, `/v1/users/${userId}`)
            authPolicy.allowMethod(HttpVerb.PATCH, `/v1/users/${userId}`)
            authPolicy.allowMethod(HttpVerb.GET, `/v1/users/${userId}/image`)
            authPolicy.allowMethod(HttpVerb.POST, `/v1/users/${userId}/image`)

            authPolicy.allowMethod(HttpVerb.GET, `/v1/users/${userId}/suggestions`)
            authPolicy.allowMethod(HttpVerb.DELETE, `/v1/users/${userId}/suggestions/*`)

            authPolicy.allowMethod(HttpVerb.GET, `/v1/users/${userId}/connections`)
            authPolicy.allowMethod(HttpVerb.GET, `/v1/users/${userId}/connections/suggested`)
            authPolicy.allowMethod(HttpVerb.GET, `/v1/users/${userId}/connections/rejected`)
            authPolicy.allowMethod(HttpVerb.GET, `/v1/users/${userId}/connections/*`)
            authPolicy.allowMethod(HttpVerb.PATCH, `/v1/users/${userId}/connections/*`)
            authPolicy.allowMethod(HttpVerb.DELETE, `/v1/users/${userId}/connections/*`)

            authPolicy.allowMethod(HttpVerb.GET, `/v1/users/${userId}/connections/requests/pending`)
            authPolicy.allowMethod(HttpVerb.POST, `/v1/users/${userId}/connections/requests`)
            authPolicy.allowMethod(HttpVerb.POST, `/v1/users/${userId}/connections/requests/confirm`)
            authPolicy.allowMethod(HttpVerb.POST, `/v1/users/${userId}/connections/requests/deny`)

            authPolicy.allowMethod(HttpVerb.POST, `/v1/users/${userId}/connection-groups`)
            authPolicy.allowMethod(HttpVerb.GET, `/v1/users/${userId}/connection-groups`)
            authPolicy.allowMethod(HttpVerb.DELETE, `/v1/users/${userId}/connection-groups/*`)
            authPolicy.allowMethod(HttpVerb.POST, `/v1/users/${userId}/connection-groups/*/export`)
            authPolicy.allowMethod(HttpVerb.GET, `/v1/users/${userId}/connection-groups/*/connections`)
            authPolicy.allowMethod(HttpVerb.POST, `/v1/users/${userId}/connection-groups/*/connections`)
            authPolicy.allowMethod(HttpVerb.DELETE, `/v1/users/${userId}/connection-groups/*/connections`)
        } else {
            authPolicy.denyAllMethods()
        }
    }
}

const authorizer = new UserAuthorizer()

export const authorize: Handler = async (event) => {
    return await authorizer.authorize(event)
}

File: apps/user/src/functions/user_authorizer.py
from authorization.abstract_authorizer import AbstractAuthorizer
from authorization.auth_policy import AuthPolicy, HttpVerb


class UserAuthorizer(AbstractAuthorizer):
    def __init__(self):
        AbstractAuthorizer.__init__(self)

    def generate_permissions(self, auth_policy: AuthPolicy, user_id: str):
        if user_id is not None:
            auth_policy.allow_method(HttpVerb.POST, f"/v1/users/{user_id}/notifications/token")
            auth_policy.allow_method(HttpVerb.GET, f"/v1/users/search")
            auth_policy.allow_method(HttpVerb.POST, f"/v1/users/search/contacts")
            auth_policy.allow_method(HttpVerb.POST, f"/v1/users/notify")
            auth_policy.allow_method(HttpVerb.GET, f"/v1/users/{user_id}")
            auth_policy.allow_method(HttpVerb.PATCH, f"/v1/users/{user_id}")
            auth_policy.allow_method(HttpVerb.DELETE, f"/v1/users/{user_id}")
            auth_policy.allow_method(HttpVerb.GET, f"/v1/users/{user_id}/image")
            auth_policy.allow_method(HttpVerb.POST, f"/v1/users/{user_id}/image")

            auth_policy.allow_method(HttpVerb.POST, f"/v1/users/{user_id}/invite")

            auth_policy.allow_method(HttpVerb.GET, f"/v1/users/{user_id}/suggestions")
            auth_policy.allow_method(HttpVerb.DELETE, f"/v1/users/{user_id}/suggestions/*")

            auth_policy.allow_method(HttpVerb.GET, f"/v1/users/{user_id}/connections")
            auth_policy.allow_method(HttpVerb.GET, f"/v1/users/{user_id}/connections/suggested")
            auth_policy.allow_method(HttpVerb.GET, f"/v1/users/{user_id}/connections/rejected")
            auth_policy.allow_method(HttpVerb.GET, f"/v1/users/{user_id}/connections/*")
            auth_policy.allow_method(HttpVerb.PATCH, f"/v1/users/{user_id}/connections/*")
            auth_policy.allow_method(HttpVerb.DELETE, f"/v1/users/{user_id}/connections/*")

            auth_policy.allow_method(HttpVerb.GET, f"/v1/users/{user_id}/connections/requests/pending")
            auth_policy.allow_method(HttpVerb.POST, f"/v1/users/{user_id}/connections/requests")
            auth_policy.allow_method(HttpVerb.POST, f"/v1/users/{user_id}/connections/requests/confirm")
            auth_policy.allow_method(HttpVerb.POST, f"/v1/users/{user_id}/connections/requests/deny")

            auth_policy.allow_method(HttpVerb.POST, f"/v1/users/{user_id}/connection-groups")
            auth_policy.allow_method(HttpVerb.GET, f"/v1/users/{user_id}/connection-groups")
            auth_policy.allow_method(HttpVerb.DELETE, f"/v1/users/{user_id}/connection-groups/*")
            auth_policy.allow_method(HttpVerb.POST, f"/v1/users/{user_id}/connection-groups/*/export")
            auth_policy.allow_method(HttpVerb.GET, f"/v1/users/{user_id}/connection-groups/*/connections")
            auth_policy.allow_method(HttpVerb.POST, f"/v1/users/{user_id}/connection-groups/*/connections")
            auth_policy.allow_method(HttpVerb.DELETE, f"/v1/users/{user_id}/connection-groups/*/connections")
        else:
            auth_policy.deny_all_methods()


authorizer = UserAuthorizer()


def authorize(event, context):
    return authorizer.authorize(event)
